File: backend/app/schemas/voice.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.job import AIJobRead


VoiceStatus = Literal[
    "no_voice",
    "default_tts",
    "sample_ready",
    "cloning",
    "cloned_ready",
    "clone_failed",
]
DefaultTTSGender = Literal["male", "female", "neutral"]
DefaultTTSAgeStyle = Literal["young", "middle_aged", "elderly"]
DefaultTTSStyle = Literal["gentle", "calm", "lively", "kind", "low"]
DefaultTTSSpeed = Literal["slow", "normal", "fast"]
DefaultTTSEmotion = Literal["calm", "comfort", "encourage", "nostalgia"]


class VoiceModelRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    persona_id: str
    provider_type: str | None
    provider_name: str | None
    status: str | None
    reference_audio_asset_id: str | None
    model_artifact_url: str | None
    sample_text: str | None
    sample_audio_url: str | None
    quality_score: int | None
    user_selected: bool
    created_at: datetime
    updated_at: datetime


class VoiceConfigResponse(BaseModel):
    persona_id: str
    voice_status: VoiceStatus
    selected_voice_model: VoiceModelRead | None
    voice_models: list[VoiceModelRead] = Field(default_factory=list)
    default_tts_notice: str
    default_tts_options: dict[str, list[str]]


class DefaultTTSSelection(BaseModel):
    gender: DefaultTTSGender = "female"
    age_style: DefaultTTSAgeStyle = "elderly"
    style: DefaultTTSStyle = "gentle"
    speed: DefaultTTSSpeed = "normal"
    emotion: DefaultTTSEmotion = "comfort"


class SpeechSynthesisCreate(BaseModel):
    text: str = Field(min_length=1)

    @field_validator("text")
    @classmethod
    def reject_blank_text(cls, value: str):
        stripped = value.strip()
        if not stripped:
            raise ValueError("text cannot be blank")
        return stripped


class SpeechSynthesisResponse(BaseModel):
    audio_url: str
    voice_status: VoiceStatus
    selected_voice_model: VoiceModelRead
    default_tts_notice: str
    provider: dict[str, Any]
    job: AIJobRead


class VoiceSampleCreate(BaseModel):
    source_material_id: str = Field(min_length=1)

    @field_validator("source_material_id")
    @classmethod
    def reject_blank_source_material_id(cls, value: str):
        stripped = value.strip()
        if not stripped:
            raise ValueError("source_material_id cannot be blank")
        return stripped


class VoiceSampleResponse(BaseModel):
    voice_status: VoiceStatus
    voice_model: VoiceModelRead
    provider: dict[str, Any]
    job: AIJobRead


class VoiceCloneCreate(BaseModel):
    voice_model_id: str | None = Field(default=None, min_length=1)
    simulate_failure: bool = False

    @field_validator("voice_model_id")
    @classmethod
    def reject_blank_voice_model_id(cls, value: str | None):
        if value is None:
            return value
        stripped = value.strip()
        if not stripped:
            raise ValueError("voice_model_id cannot be blank")
        return stripped


class VoiceCloneResponse(BaseModel):
    voice_status: VoiceStatus
    voice_model: VoiceModelRead
    selected_voice_model: VoiceModelRead
    default_tts_notice: str
    provider: dict[str, Any]
    job: AIJobRead

File: frontend/src/lib/voice.ts
import { authHeaders } from "./auth";
import { API_PATHS, buildApiUrl, readApiJson } from "./api";

export const DEFAULT_TTS_NOTICE =
  "当前使用系统默认声音，不是 TA 的真实声音。上传 TA 的清晰语音后，可以尝试生成模拟音色。";

export type VoiceStatus =
  | "no_voice"
  | "default_tts"
  | "sample_ready"
  | "cloning"
  | "cloned_ready"
  | "clone_failed";

export type VoiceModelRead = {
  id: string;
  persona_id: string;
  provider_type: string | null;
  provider_name: string | null;
  status: string | null;
  reference_audio_asset_id: string | null;
  model_artifact_url: string | null;
  sample_text: string | null;
  sample_audio_url: string | null;
  quality_score: number | null;
  user_selected: boolean;
  created_at: string;
  updated_at: string;
};

export type VoiceConfigResponse = {
  persona_id: string;
  voice_status: VoiceStatus;
  selected_voice_model: VoiceModelRead | null;
  voice_models: VoiceModelRead[];
  default_tts_notice: string;
  default_tts_options: Record<string, string[]>;
};

export type DefaultTTSPayload = {
  gender: "male" | "female" | "neutral";
  age_style: "young" | "middle_aged" | "elderly";
  style: "gentle" | "calm" | "lively" | "kind" | "low";
  speed: "slow" | "normal" | "fast";
  emotion: "calm" | "comfort" | "encourage" | "nostalgia";
};

export type VoiceJobRead = {
  id: string;
  job_type: string;
  status: string;
};

export type SpeechSynthesisResponse = {
  audio_url: string;
  voice_status: VoiceStatus;
  selected_voice_model: VoiceModelRead;
  default_tts_notice: string;
  job: VoiceJobRead;
};

export type VoiceSampleResponse = {
  voice_status: VoiceStatus;
  voice_model: VoiceModelRead;
  job: VoiceJobRead;
};

export type VoiceCloneResponse = {
  voice_status: VoiceStatus;
  voice_model: VoiceModelRead;
  selected_voice_model: VoiceModelRead;
  default_tts_notice: string;
  job: VoiceJobRead;
};

export const VOICE_API_PATHS = API_PATHS.voice;

export function personaVoiceRoute(personaId: string): string {
  return `/personas/${encodeURIComponent(personaId)}/voice`;
}

export function buildDefaultTtsPayload(): DefaultTTSPayload {
  return {
    gender: "female",
    age_style: "elderly",
    style: "gentle",
    speed: "normal",
    emotion: "comfort"
  };
}

export function voiceStatusLabel(status: string | null | undefined): string {
  switch (status) {
    case "no_voice":
      return "未设置声音";
    case "default_tts":
      return "系统默认 TTS";
    case "sample_ready":
      return "已有可克隆样本";
    case "cloning":
      return "音色克隆中";
    case "cloned_ready":
      return "模拟音色可用";
    case "clone_failed":
      return "音色克隆失败";
    default:
      return status || "未知状态";
  }
}

export function isBlankVoiceText(text: string): boolean {
  return text.trim().length === 0;
}

export function voiceModelSummary(model: VoiceModelRead): string {
  const parts = [
    voiceStatusLabel(model.status),
    model.provider_name,
    model.quality_score === null ? null : `质量 ${model.quality_score}`
  ].filter(Boolean);
  return parts.join(" · ");
}

export async function getVoiceConfig(personaId: string): Promise<VoiceConfigResponse> {
  const response = await fetch(buildApiUrl(VOICE_API_PATHS.config(personaId)), {
    headers: authHeaders()
  });
  return readApiJson<VoiceConfigResponse>(response, "无法加载声音配置。");
}

export async function selectDefaultTts(
  personaId: string,
  payload: DefaultTTSPayload = buildDefaultTtsPayload()
): Promise<VoiceConfigResponse> {
  const response = await fetch(buildApiUrl(VOICE_API_PATHS.defaultTts(personaId)), {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify(payload)
  });
  return readApiJson<VoiceConfigResponse>(response, "无法选择默认 TTS。");
}

export async function createVoiceSample(
  personaId: string,
  sourceMaterialId: string
): Promise<VoiceSampleResponse> {
  const response = await fetch(buildApiUrl(VOICE_API_PATHS.samples(personaId)), {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify({ source_material_id: sourceMaterialId })
  });
  return readApiJson<VoiceSampleResponse>(response, "无法创建音色样本。");
}

export async function cloneVoice(
  personaId: string,
  voiceModelId: string
): Promise<VoiceCloneResponse> {
  const response = await fetch(buildApiUrl(VOICE_API_PATHS.clone(personaId)), {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify({ voice_model_id: voiceModelId })
  });
  return readApiJson<VoiceCloneResponse>(response, "无法发起音色克隆。");
}

export async function synthesizeSpeech(
  personaId: string,
  text: string
): Promise<SpeechSynthesisResponse> {
  const response = await fetch(buildApiUrl(VOICE_API_PATHS.synthesize(personaId)), {
    method: "POST",
    headers: { "Content-Type": "application/json", ...authHeaders() },
    body: JSON.stringify({ text })
  });
  return readApiJson<SpeechSynthesisResponse>(response, "无法生成语音。");
}
